src/libponyc/expr/lambda.c

```diff
@@ -182,3 +185,341 @@ bool expr_lambda(pass_opt_t* opt, ast_t** astp)
 
   return ast_passes_subtree(astp, opt, PASS_EXPR);
 }
+
+
+static bool capture_from_reference(pass_opt_t* opt, ast_t* ctx, ast_t* ast,
+  ast_t* captures, ast_t** last_capture)
+{
+  const char* name = ast_name(ast_child(ast));
+
+  sym_status_t status;
+  ast_t* refdef = ast_get(ast, name, &status);
+
+  if(refdef != NULL)
+    return true;
+
+  refdef = ast_get(ctx, name, &status);
+
+  if(refdef == NULL)
+  {
+    ast_error(opt->check.errors, ast,
+      "cannot capture \"%s\", variable not defined", name);
+    return false;
+  }
+
+  if(!def_before_use(opt, refdef, ctx, name))
+    return false;
+
+  token_id def_id = ast_id(refdef);
+
+  if(def_id != TK_ID && def_id != TK_FVAR && def_id != TK_FLET &&
+    def_id != TK_PARAM)
+  {
+    ast_error(opt->check.errors, ast, "cannot capture \"%s\", can only "
+      "capture fields, parameters and local variables", name);
+    return false;
+  }
+
+  // Check if we've already captured it
+  for(ast_t* p = ast_child(captures); p != NULL; p = ast_sibling(p))
+  {
+    AST_GET_CHILDREN(p, c_name, c_type);
+
+    if(name == ast_name(c_name))
+      return true;
+  }
+
+  // TODO: add as a field, add as a constructor parameter, init field from
+  // parameter in constructor, add at the call site
```

Comment:
does this TODO need to be handled before merging? Should we create an issue for it?